### ISSUE TITLE
Removing packages from MELPA

### ISSUE THREAD

**Issue by phikal** [melpa/melpa#7305]
Hi,

I was thinking about removing [packages I have contributed from MELPA](https://github.com/melpa/melpa/commits?author=phikal), and moving them all to GNU ELPA, mainly because I don't want to deal with issues stemming from every commit being turned into a new package, and that I don't want to burden users with configuring MELPA, just to use a package.

The reason I opened this issue is because I couldn't find any previous discussions on this topic, and I'm not sure what the maintainers policy is, as they are all free software, and legaly can be distributed by whomever.

**Comment by alphapapa** [comment]
I think MELPA will generally respect an author's wishes, so if you want them removed, they probably will be.  If a package had many users, and removing it from MELPA would inconvenience many people, I suppose they might decline to stop distributing them, but they might agree to stop updating them, and/or to remove them eventually, after users had a chance to install from other sources.

However, the two issues you point out can be solved by other means.  For example, instead of pushing untested commits to `master`, push them to a branch, and only push to `master` when ready for users.  And if you don't want users to have to configure MELPA, offer alternative installation instructions.  For example, you could show them how to use Quelpa, which installs directly from package repos: https://github.com/alphapapa/bufler.el#quelpa

**Comment by phikal** [comment]
> If a package had many users, and removing it from MELPA would inconvenience many people

Looking at the current download numbers from melpa.org:

| Package | # MELPA | # MELPA Stable | Total |
|-|-|-|-|
| 0x0 | 2039 | 120 | 2159 |
| clang-capf | 558 | 78 |  636 |
| go-capf | 482 | 78 | 560 |
| face-shift | 70 | 20 | 90 | |
| insert-kaomoji | 169 | 28 | 197  |
| bang | 707 | 59 | 766 |
| autocrypt | 168 | - | 168 |

The only package I would be worried about is 0x0, but others like go-capf are basically deprecated because [gocode](https://github.com/nsf/gocode) has been abandoned. bang is also not being worked on any more, since it's renamed version has been added to [ELPA](http://elpa.gnu.org/packages/shell-command+.html).

On a related point: I'm not familiar enough with the internals of package.el, but is there a way for Emacs to pick up that a package has switched archives? If not, this should be fixed before anything is changed, specially considering that with the upcoming NonGNU ELPA, a lot more people won't be needing MELPA anymore.

> And if you don't want users to have to configure MELPA, offer alternative installation instructions. For example, you could show them how to use Quelpa, which installs directly from package repos: https://github.com/alphapapa/bufler.el#quelpa

That doesn't make it much better. The main advantange of the ELPAs are that they work OOTB, and this seems to be another extention, that just makes the lives of new users more complicated.

**Comment by alphapapa** [comment]
> with the upcoming NonGNU ELPA, a lot more people won't be needing MELPA anymore.

I've been following its developments on emacs-devel, and while I'm hopeful that it turns out to be useful, it remains to be seen what it will amount to.  It was only a few days ago that people were asking (again) what the nonGNU ELPA's purpose is intended to be.  They have yet to decide how packages will be added, which packages will be added, what the criteria will be, how authors will propagate updates, etc.  It doesn't seem that anyone has a concrete vision for what it is supposed to be and how it's supposed to work.  

Meanwhile, I just published another package to MELPA today, with minimal hassle and delay, and users are already installing it.  I think MELPA's value is self-evident.

**Comment by phikal** [comment]
It goes without saying that the package archive that currently exists has an advantage over that one that doesn't exist, but I didn't open this issue to debate MELPA vs NonGNU ELPA: It's obvious we disagree on this.

What I want to find out is what the best transition-plan is, for packages that do want to move from MELPA to NonGNU ELPA or in my case GNU ELPA.

**Comment by purcell** [comment]
You're entirely welcome to switch your packages to ELPA, and people have done it in the past. But people have also switched back, or chosen to host in both archives (mmm-mode comes to mind).

**Comment by purcell** [comment]
> What I want to find out is what the best transition-plan is, for packages that do want to move from MELPA to NonGNU ELPA or in my case GNU ELPA.

1. Set up hosting in new archive
2. Submit PR to remove recipes from MELPA.

**Comment by purcell** [comment]
Oh, and for anything that's basically defunct and unusable (go-capf?), we're happy to just remove it.

**Comment by phikal** [comment]
> 1. Set up hosting in new archive
> 2. Submit PR to remove recipes from MELPA.

So all in all, there is no way to make the transition from one package repository to another easy/unnoticed for users?

**Comment by alphapapa** [comment]
AFAIK the issue is the package version numbers.  MELPA sets the package versions to timestamp-based numbers, which Emacs considers newer than any normal version number.  So in order for a user to receive updates to a package from a non-MELPA repo, the package would have to be deleted and reinstalled from the non-MELPA repo first.

You could try releasing a final update to the MELPA version of the package and adding a warning message instructing the user to delete the package and install it from the other repo.

**Comment by purcell** [comment]
> > 2\.
> 
> So all in all, there is no way to make the transition from one package repository to another easy/unnoticed for users?

Well, no, at some point you will have physically removed it from the place users were previously getting it. On a long enough timescale there will always be users caught out by its apparent disappearance.

**Comment by basil-conto** [comment]
> So in order for a user to receive updates to a package from a non-MELPA repo, the package would have to be deleted and reinstalled from the non-MELPA repo first.

As well as set `package-archives`, `package-archive-priorities`, and/or `package-pinned-packages` appropriately depending on Emacs version + existing user configuration.

**Comment by tarsius** [comment]
Let us know if you have any more questions.

**Comment by tarsius** [comment]
What happened to `0x0`?

I cannot find it on GNU ELPA or NonGNU ELPA, and https://gitlab.com/willvaughn/emacs-0x0 has also disappeared. A copy is available on the Emacsmirror. But something seems to have gone wrong on my end; on the machine I use to update the mirror, the repository is also missing. I would have to manually restore it from https://github.com/emacsmirror/0x0 before moving it from the mirror to the attic, but maybe I can avoid that by mirroring it from the new upstream location instead?

/cc @nackjicholson

**Comment by phikal** [comment]
If nothing happens, I can take over maintenance again.

**Comment by phikal** [comment]
FWIW I have brought my fork up to date with William's, but if I were to take over again, I'd prefer adding the package to GNU ELPA, so I'd also remove it from MELPA as well.

**Comment by nackjicholson** [comment]
@tarsius @phikal I moved it to sourcehut, I believe I put a PR in to MELPA to do that.

**Comment by nackjicholson** [comment]
@tarsius @phikal https://github.com/melpa/melpa/pull/8566

Did I not do something correctly?

**Comment by tarsius** [comment]
I got confused because you did not update this header: `;; Homepage: https://gitlab.com/willvaughn/emacs-0x0`. And my tools messed up mirroring from the new location, but that is most likely completely unrelated and just a coincidence.

In other words, aside from the outdated information in the header, everything seems to be fine. I have just updated the Emacsmirror to mirror from the new location and that worked without any issues.
